### PR TITLE
Add Reformer Model

### PR DESCRIPTION
**Pull request by JunnYu**
飞桨论文复现挑战赛（第四期）Reformer: The Efficient Transformer 论文复现提交。

### CODE REVIEW

**Comment by yingyibiao** [comment]
前向输出还需要再定位清楚不一致的op！

File: paddlenlp/transformers/reformer/modeling.py [hunk 1]
```diff
@@ -0,0 +1,2755 @@
...
+)
+ReformerEncoderOutput = namedtuple(
+    "ReformerEncoderOutput",
+    [
+        "hidden_states", "all_hidden_states", "all_attentions",
+        "past_buckets_states"
+    ], )
+
+
+def _logsumexp(x, axis=-1, keepdim=False):
```
Comment (joey12300) [inline paddlenlp/transformers/reformer/modeling.py:62]: 是否可以解释一下为何要二次封装logsumexp呢？如果原有的paddle.logsumexp可以解决问题的话没有必要再封装一层，会影响性能

File: paddlenlp/transformers/reformer/modeling.py [hunk 2]
```diff
@@ -0,0 +1,2755 @@
...
+                    dtype=paddle.get_default_dtype(),
+                    default_initializer=nn.initializer.Constant(value=1.0), ))
+
+    def forward(self, position_ids):
+        # broadcast weights to correct shape
+        batch_size = position_ids.shape[0]
+        sequence_length = position_ids.shape[1]
+
+        broadcasted_weights = [
+            weight.expand(shape=(batch_size, ) + tuple(self.axial_pos_shape) +
```
Comment (joey12300) [inline paddlenlp/transformers/reformer/modeling.py]: 这里是否需要加tuple做强制转换呢？本身是否就是tuple?

File: paddlenlp/transformers/reformer/modeling.py [hunk 3]
```diff
@@ -0,0 +1,2755 @@
...
+
+        else:
+            if reduce(mul, self.axial_pos_shape) < sequence_length:
+                raise ValueError(
+                    f"Make sure that pd_config.axial_pos_shape factors: {self.axial_pos_shape} multiply at least to "
+                    f"max(sequence_length, least_common_mult_chunk_length): max({sequence_length}, "
+                    f"{self.least_common_mult_chunk_length}).")
+
+            # compute how many columns are needed
+            max_position_id = position_ids.max().item()
```
Comment (joey12300) [inline paddlenlp/transformers/reformer/modeling.py:430]: 这里尝试去掉item()，本质上max_position_id在后面代码是作为下标使用，理论上无需使用item()取值也是可以运行的。在训练代码加入item()这种函数会导致无法转成静态图的部署代码

File: paddlenlp/transformers/reformer/modeling.py [hunk 4]
```diff
@@ -0,0 +1,2755 @@
...
+                if buckets is None:
+                    buckets = paddle.argmax(rotated_vectors_factor, axis=-1)
+                else:
+                    buckets = buckets + (cur_product * paddle.argmax(
+                        rotated_vectors_factor, axis=-1))
+
+                cur_product = cur_product * bucket_factor
+
+        if attention_mask is not None and (
+                attention_mask.sum().item() < batch_size *
```
Comment (joey12300) [inline paddlenlp/transformers/reformer/modeling.py]: 同上，理论上也是可以去掉item()

File: paddlenlp/transformers/reformer/modeling.py [hunk 5]
```diff
@@ -0,0 +1,2755 @@
...
+        # randomize seeds
+        # use cuda generator if available
+        if paddle.get_device() != "cpu":
+            # GPU
+            device_idx = int(paddle.get_device().split(":")[1])
+            sts = paddle.get_cuda_rng_state()
+            self.attention_seed = sts[device_idx].current_seed()
+        else:
+            # CPU
+            self.attention_seed = int(999999999 % sys.maxsize)
```
Comment (joey12300) [inline paddlenlp/transformers/reformer/modeling.py]: 这里CPU的seed为何使用固定数值呢？

File: paddlenlp/transformers/reformer/modeling.py [hunk 6]
```diff
@@ -0,0 +1,2755 @@
...
+        # randomize seeds
+        # use cuda generator if available
+        if paddle.get_device() != "cpu":
+            # GPU
+            device_idx = int(paddle.get_device().split(":")[1])
+            sts = paddle.get_cuda_rng_state()
+            self.feed_forward_seed = sts[device_idx].current_seed()
+        else:
+            # CPU
+            self.feed_forward_seed = int(999999999 % sys.maxsize)
```
Comment (joey12300) [inline paddlenlp/transformers/reformer/modeling.py]: 同上

**Review by joey12300 — CHANGES_REQUESTED** [review]

File: paddlenlp/transformers/reformer/modeling.py (hunk 1 above)
Comment (JunnYu) [reply paddlenlp/transformers/reformer/modeling.py:62]: `paddle.logsumexp` 不支持 5D的Tensor，因此我reshape成4D，再使用。

File: paddlenlp/transformers/reformer/modeling.py (hunk 2 above)
Comment (JunnYu) [reply paddlenlp/transformers/reformer/modeling.py]: 要么都使用list，要么都使用tuple。其中`self.axial_pos_shape` 是 list类型，`x.shape`返回的是list类型

File: paddlenlp/transformers/reformer/modeling.py (hunk 5 above)
Comment (JunnYu) [reply paddlenlp/transformers/reformer/modeling.py]: 我当时应该是没找到哪个API能够获取 cpu环境下的seed 。pytorch可以这样获取，torch.seed()

File: paddlenlp/transformers/reformer/modeling.py (hunk 2 above)
Comment (joey12300) [reply paddlenlp/transformers/reformer/modeling.py]: 嗯，了解。那么可以都使用list，只要shape=[batch_size, ]，后面就无需强制转换了

File: paddlenlp/transformers/reformer/modeling.py (hunk 1 above)
Comment (joey12300) [reply paddlenlp/transformers/reformer/modeling.py:62]: 了解

File: paddlenlp/transformers/reformer/modeling.py (hunk 5 above)
Comment (joey12300) [reply paddlenlp/transformers/reformer/modeling.py]: 是否可以使用原生random库或者numpy的random模块获取seed呢？

Comment (JunnYu) [reply paddlenlp/transformers/reformer/modeling.py]: 当前使用了np.random.randint

File: paddlenlp/transformers/reformer/modeling.py (hunk 4 above)
Comment (JunnYu) [reply paddlenlp/transformers/reformer/modeling.py]: 去掉了

File: paddlenlp/transformers/reformer/modeling.py (hunk 3 above)
Comment (JunnYu) [reply paddlenlp/transformers/reformer/modeling.py:430]: `无法转成静态图的部署代码`，我发现当前很多动态图代码都无法转成静态图的部署代码。因为使用xxx.shape时候，可能某些shape为None，是动态变化的。

**Comment by JunnYu** [comment]
模型权重链接。
链接：https://pan.baidu.com/s/1ZUxfged0lUavaCBb_uSZHw 
提取码：byg9 
--来自百度网盘超级会员V5的分享

**Review by joey12300 — DISMISSED** [review]
LGTM

File: paddlenlp/transformers/reformer/tokenizer.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            "reformer-crime-and-punishment":
+            "http://paddlenlp.bj.bcebos.com/models/transformers/reformer/reformer-crime-and-punishment/spiece.model",
+        },
+    }
+
+    pretrained_init_configuration = {
+        "reformer-crime-and-punishment": {
+            "do_lower_case": False
+        },
+    }
```
Comment (yingyibiao) [inline paddlenlp/transformers/reformer/tokenizer.py:63]: 这里缺少reformer-enwik8权重加载的tokenizer

**Comment by yingyibiao** [comment]
权重已上传

File: paddlenlp/transformers/reformer/tokenizer.py (hunk 7 above)
Comment (JunnYu) [reply paddlenlp/transformers/reformer/tokenizer.py:63]: reformer-enwik8 不使用`tokenizer`

**Review by yingyibiao — APPROVED** [review]
LGTM
